COMDOX-775: Update the Metadata documentation in mftf dev docs (#63)
AdobeDocs/commerce-testing: src/pages/functional-testing-framework/metadata.md

As a result, the Functional Testing Framework sends an unauthorized POST request with an empty body to the `https://example.com/rest/V1/guest-carts` and stores the single string response that the endpoint returns.

> #### Company Relation Example

Reviewer: #### `Creating a relation between two companies Example` would be clear

Author: this is the main heading before the create and delete heading

> This is an example of how to handle a company relation using REST API operations. This functionality is available only to Adobe Commerce customers.

Author: fixed

> This is an example of how to handle a company relation using REST API operations. This functionality is available only to Adobe Commerce customers.

Reviewer 2: ```suggestion
This is an example of how to handle a company relation using REST API operations. This functionality is available only to Adobe Commerce B2B customers.
```

_CompanyRelations/Data/CompanyIdsData.xml_:

```xml
<entity name="CompanyRelationData" type="company_relation">
    <var key="company_id" entityType="company2" entityKey="id"/>
</entity>
```

- `type="company_relation"` points to the operation with `dataType=company_relation"` and `type="create"` in the _Metadata_ directory.
- The `company_id` will only get the `id` from the company with `type="company2"`

_CompanyRelation/Metadata/CompanyIdsMeta.xml_:

```xml
<operation name="CreateCompanyRelation" dataType="company_relation" type="create" >
    <field key="company_id">integer</field>
</operation>
```

`<field key="company_id">integer</field>` will create an object with a key `company_id` with a value of type integer.
```json
{
  "company_id": 1
}
```

_CompanyRelation/Metadata/CompanyRelationsMeta.xml_:

```xml
<operation name="CreateCompanyRelations" dataType="company_relations" type="create" auth="adminOauth" url="/V1/company/{company.id}/relations" method="POST">
    <contentType>application/json</contentType>
    <array key="relations">
        <value>company_relation</value>
    </array>
</operation>
```

> As a result, the Functional Testing Framework sends a POST request with an array of company_ids in the body to the `https://example.com/rest/V1/company/:parentId/relations`.

Reviewer 3: > https://example.com/rest/V1/company/:parentId/relations

change to "https://example.com/rest/V1/company/{company.id}/relations"

Author: fixed

> As a result, the Functional Testing Framework sends a POST request with an array of company_ids in the body to the `https://example.com/rest/V1/company/:parentId/relations`.

Reviewer 3: should we leave a note explaining that there is no way to send more than 1 company_relation to the array?

Reviewer: Maybe say that Currently, the MFTF supports sending only one company as a child company for a parent.

- `{company.id}` in the url comes from the test `<requiredEntity createDataKey="company1"/>` in EstablishOneCompanyRelationData that has `type="company"`
- `<array key="relations">` will create an array with the key relations
- `<value>company_relation</value>` will get the data from the operation with `dataType="company_relation"`
```json
{
  "relations": [
    {
      "company_id":  1
    }
  ]
}
```

##### Delete a company relation

```xml
<deleteData createDataKey="createCompanyRelation" stepKey="deleteCompanyRelation"/>
```

> `createDataKey="createCompanyRelation"` will be the same as the createData `stepKey="createCompanyRelation"` in the test. This will call the delete operation that has the same `dataType="company_relations"` as the createData.

Author: fixed

```xml
<operation name="DeleteCompanyRelations" dataType="company_relations" type="delete" auth="adminOauth" url="/V1/company/{company.id}/relations/{company2.id}" method="DELETE">
    <contentType>application/json</contentType>
</operation>
```

> As a result, the Functional Testing Framework sends a DELETE request to the `https://example.com/rest/V1/company/:parentId/relations/:companyId`.

Reviewer 3: > https://example.com/rest/V1/company/:parentId/relations/:companyId

change to "https://example.com/rest/V1/company/{company.id}/relations/{company2.id}"

Author: fixed

#### Company Relation Example

This is an example of how to create the _Data_ and _Metadata_ that allows the Functional Testing Framework to process the REST API operations for creating and deleting a company relation. This functionality is available only to Adobe Commerce B2B customers.

REST API endpoints:

- POST: `/V1/company/:parentId/relations`
- DELETE: `/V1/company/:parentId/relations/:companyId`

##### Create a company relation

The corresponding test step is:

First create 2 different companies of different types before creating the company relation

```xml
<createData entity="Simple_US_Customer" stepKey="companyAdmin1"/>
<createData entity="ParentCompany" stepKey="parentCompany">
    <requiredEntity createDataKey="companyAdmin1"/>
</createData>

<createData entity="Simple_US_CA_Customer" stepKey="companyAdmin2"/>
<createData entity="ChildCompany" stepKey="childCompany">
    <requiredEntity createDataKey="companyAdmin2"/>
</createData>

<createData entity="OneCompanyRelation" stepKey="createCompanyRelation">
    <requiredEntity createDataKey="parentCompany"/>
    <requiredEntity createDataKey="childCompany"/>
</createData>
```

> The Functional Testing Framework searches in the _Data_ directory for the entities with `<entity name="Default_Company">`, `<entity name="NewCompany">`, and `<entity name="CreateCompanyRelationData">` and reads `type`.

Reviewer 3: need to update all 3 entity names

    <data key="company_name" unique="suffix">Parent Company </data>
</entity>
<!-- Data Type "company2" -->
<entity name="ChildCompany" type="company2">
    <data key="status">1</data>
    <data key="company_name" unique="suffix">Child Company </data>
    <array key="street">
        <item>7700 W Parmer Ln</item>
        <item>Bld D</item>
    </array>
    <data key="city">Culver City</data>
    <data key="country_id">US</data>
    <data key="region_id">57</data>
    <data key="postcode">90230</data>
    <data key="telephone">555-555-5555</data>
    <data key="customer_group_id">1</data>
    <data key="sales_representative_id">1</data>
    <var key="super_user_id" entityType="customer" entityKey="id"/>
    <var key="company_email" entityType="customer" entityKey="email"/>
</entity>
```

Notice that the 2 company entities have 2 different types `type="company"` and `type="company2"`.  These types will be used later in _Data_ fields and _Metadata_ urls for company relations.

_CompanyRelation/Data/CompanyRelationData.xml_:

```xml
<entity name="OneCompanyRelation" type="company_relation">
    <requiredEntity type="company_id">CompanyId</requiredEntity>
</entity>
```

> - `type="company_relations"` points to the operation with `dataType=company_relations"` and `type="create"` in the _Metadata_ directory.

Reviewer 3: update type & dataType

> - `<requiredEntity type="company_relation">CompanyRelationData</requiredEntity>` points to the data entity with `name="CompanyRelationData"` and `type="company_relation"`.

Reviewer 3: update referenced data names & type

_CompanyRelations/Data/CompanyIdData.xml_:

```xml
<entity name="CompanyId" type="company_id">
    <var key="company_id" entityType="company2" entityKey="id"/>
</entity>
```

> - `type="company_relation"` points to the operation with `dataType=company_relation"` and `type="create"` in the _Metadata_ directory.

Reviewer 3: update type & dataType

```xml
<operation name="CreateCompanyId" dataType="company_id" type="create">
    <field key="company_id">integer</field>
</operation>
```

`<field key="company_id">integer</field>` will create an object with a key `company_id` with a value of type integer.

```json
{
  "company_id": 1
}
```

_CompanyRelation/Metadata/CompanyRelationMeta.xml_:

```xml
<operation name="CreateCompanyRelation" dataType="company_relation" type="create" auth="adminOauth" url="/V1/company/{company.id}/relations" method="POST">
    <contentType>application/json</contentType>
    <array key="relations">
        <value>company_id</value>
    </array>
</operation>
```

As a result, the Functional Testing Framework sends a POST request with an array of company_ids in the body to the `https://example.com/rest/V1/company/{company.id}/relations`. Currently, the Functional Testing Framework supports sending only one company as a child company for a parent.

- `{company.id}` in the url comes from the test `<requiredEntity createDataKey="company1"/>` in EstablishOneCompanyRelationData that has `type="company"`
- `<array key="relations">` will create an array with the key relations

> - `<value>company_relation</value>` will get the data from the operation with `dataType="company_relation"`

Reviewer 3: update to company_id

```json
{
  "relations": [
    {
      "company_id":  1
    }
  ]
}
```

##### Delete a company relation

```xml
<deleteData createDataKey="createCompanyRelation" stepKey="deleteCompanyRelation"/>
```

> `createDataKey="createCompanyRelation"` will be the same as the createData with `stepKey="createCompanyRelation"` in the test. This will call the delete operation that has the same `dataType="company_relations"` as the createData.

Reviewer 3: update dataType

_CompanyRelation/Metadata/CompanyRelationMeta.xml_:

```xml
<operation name="DeleteCompanyRelation" dataType="company_relation" type="delete" auth="adminOauth" url="/V1/company/{company.id}/relations/{company2.id}" method="DELETE">
    <contentType>application/json</contentType>
</operation>
```

As a result, the Functional Testing Framework sends a DELETE request to the `https://example.com/rest/V1/company/{company.id}/relations/{company2.id}`.

> - `{company.id}` in the url comes from the test `<requiredEntity createDataKey="company1"/>` in EstablishOneCompanyRelationData that has `type="company"`

Reviewer 3: update EstablishOneCompanyRelationData

> - `{company2.id}` in the url comes from the test `<requiredEntity createDataKey="company2"/>` in EstablishOneCompanyRelationData that has `type="company2"`

Reviewer 3: update EstablishOneCompanyRelationData

As a result, the Functional Testing Framework sends an unauthorized POST request with an empty body to the `https://example.com/rest/V1/guest-carts` and stores the single string response that the endpoint returns.

#### Company Relation Example

> This is an example of how to create the _Data_ and _Metadata_ that allows the Functional Testing Framework to process the REST API operations for creating and deleting a company relation. This functionality is available only to Adobe Commerce B2B customers.

Reviewer 2: ```suggestion
This example creates XML data files in the_Data_ and _Metadata_ directories that allow the Functional Testing Framework to process the REST API operations for creating and deleting a company relation. This functionality is available only to Adobe Commerce B2B customers.
```

REST API endpoints:

- POST: `/V1/company/:parentId/relations`
- DELETE: `/V1/company/:parentId/relations/:companyId`

##### Create a company relation

The corresponding test step is:

First create 2 different companies of different types before creating the company relation

```xml
<createData entity="Simple_US_Customer" stepKey="companyAdmin1"/>
<createData entity="ParentCompany" stepKey="parentCompany">
    <requiredEntity createDataKey="companyAdmin1"/>
</createData>

<createData entity="Simple_US_CA_Customer" stepKey="companyAdmin2"/>
<createData entity="ChildCompany" stepKey="childCompany">
    <requiredEntity createDataKey="companyAdmin2"/>
</createData>

<createData entity="OneCompanyRelation" stepKey="createCompanyRelation">
    <requiredEntity createDataKey="parentCompany"/>
    <requiredEntity createDataKey="childCompany"/>
</createData>
```

> The Functional Testing Framework searches in the _Data_ directory for the entities with `<entity name="ParentCompany">`, `<entity name="ChildCompany">`, and `<entity name="OneCompanyRelation">` and reads `type`.

Reviewer 2: ```suggestion
The Functional Testing Framework searches in the _Data_ directory for the entities with `<entity name="ParentCompany">`, `<entity name="ChildCompany">`, and `<entity name="OneCompanyRelation">` and reads the `type`.
```

    <data key="sales_representative_id">1</data>
    <var key="super_user_id" entityType="customer" entityKey="id"/>
    <var key="company_email" entityType="customer" entityKey="email"/>
</entity>
```

Notice that the 2 company entities have 2 different types `type="company"` and `type="company2"`.  These types will be used later in _Data_ fields and _Metadata_ urls for company relations.

_CompanyRelation/Data/CompanyRelationData.xml_:

```xml
<entity name="OneCompanyRelation" type="company_relation">
    <requiredEntity type="company_id">CompanyId</requiredEntity>
</entity>
```

- `type="company_relation"` points to the operation with `dataType=company_relation"` and `type="create"` in the _Metadata_ directory.
- `<requiredEntity type="company_id">CompanyId</requiredEntity>` points to the data entity with `name="CompanyId"` and `type="company_id"`.

_CompanyRelations/Data/CompanyIdData.xml_:

```xml
<entity name="CompanyId" type="company_id">
    <var key="company_id" entityType="company2" entityKey="id"/>
</entity>
```

- `type="company_id"` points to the operation with `dataType=company_id"` and `type="create"` in the _Metadata_ directory.

> - The `company_id` will only get the `id` from the company with `type="company2"`

Reviewer 2: ```suggestion
- The `company_id` retrieves only  the `id` from the company with `type="company2"`
```

_CompanyRelation/Metadata/CompanyIdMeta.xml_:

```xml
<operation name="CreateCompanyId" dataType="company_id" type="create">
    <field key="company_id">integer</field>
</operation>
```

> `<field key="company_id">integer</field>` will create an object with a key `company_id` with a value of type integer.

Reviewer 2: ```suggestion
`<field key="company_id">integer</field>` creates an object with a key `company_id` with a value of type integer.
```

```json
{
  "company_id": 1
}
```

_CompanyRelation/Metadata/CompanyRelationMeta.xml_:

```xml
<operation name="CreateCompanyRelation" dataType="company_relation" type="create" auth="adminOauth" url="/V1/company/{company.id}/relations" method="POST">
    <contentType>application/json</contentType>
    <array key="relations">
        <value>company_id</value>
    </array>
</operation>
```

> As a result, the Functional Testing Framework sends a POST request with an array of company_ids in the body to the `https://example.com/rest/V1/company/{company.id}/relations`. Currently, the Functional Testing Framework supports sending only one company as a child company for a parent.

Reviewer 2: ```suggestion
As a result, the Functional Testing Framework sends a POST request with an array of `company_ids` in the body to `https://example.com/rest/V1/company/{company.id}/relations`. Currently, the Functional Testing Framework supports sending only one company as a child company for a parent.
```

- `{company.id}` in the url comes from the test `<requiredEntity createDataKey="parentCompany"/>` in OneCompanyRelation that has `type="company"`

> - `<array key="relations">` will create an array with the key relations
> - `<value>company_id</value>` will get the data from the operation with `dataType="company_id"`

Reviewer 2: ```suggestion
- `<array key="relations">` creates an array with the key relations
- `<value>company_id</value>` retrieves the data from the operation with `dataType="company_id"`
```

```json
{
  "relations": [
    {
      "company_id":  1
    }
  ]
}
```

##### Delete a company relation

The corresponding test step is:

```xml
<deleteData createDataKey="createCompanyRelation" stepKey="deleteCompanyRelation"/>
```

> `createDataKey="createCompanyRelation"` will be the same as the createData with `stepKey="createCompanyRelation"` in the test. This will call the delete operation that has the same `dataType="company_relation"` as the createData.

Reviewer 2: ```suggestion
`createDataKey="createCompanyRelation"` is the same as the `createData` entity with `stepKey="createCompanyRelation"` in the test. This step calls the delete operation that has the same `dataType="company_relation"` value as the `createData` entity.
```

_CompanyRelation/Metadata/CompanyRelationMeta.xml_:

```xml
<operation name="DeleteCompanyRelation" dataType="company_relation" type="delete" auth="adminOauth" url="/V1/company/{company.id}/relations/{company2.id}" method="DELETE">
    <contentType>application/json</contentType>
</operation>
```

> As a result, the Functional Testing Framework sends a DELETE request to the `https://example.com/rest/V1/company/{company.id}/relations/{company2.id}`.

Reviewer 4: ```suggestion
As a result, the Functional Testing Framework sends a DELETE request to `https://example.com/rest/V1/company/{company.id}/relations/{company2.id}`.
```

Reviewer 4: any prefix of "the" where URL literal acts as the noun is slightly jarring.  really really really minor

> - `{company.id}` in the url comes from the test `<requiredEntity createDataKey="parentCompany"/>` in OneCompanyRelation that has `type="company"`
> - `{company2.id}` in the url comes from the test `<requiredEntity createDataKey="childCompany"/>` in OneCompanyRelation that has `type="company2"`

Reviewer 2: ```suggestion
- `{company.id}` in the url comes from the test `<requiredEntity createDataKey="parentCompany"/>` in the `OneCompanyRelation` entity that has `type="company"`
- `{company2.id}` in the url comes from the test `<requiredEntity createDataKey="childCompany"/>` in the `OneCompanyRelation` entity that has `type="company2"`
```

### Handling a REST API response

> There are cases when you need to reuse the data that Adobe Commerce or Magento Open source responded with to your POST request.

Reviewer 2: ```suggestion
There are cases when you need to reuse the data that Adobe Commerce or Magento Open source returned in response to your POST request.
```
